## EJERCICIO 1 - BASICO

El fichero datos_pago_microcreditos.csv contiene 30000 transacciones bancarias relativas al pago o impago de microcreditos. El dataset consta de 62 dimensiones propietarias de las cuales el banco no ha proporcionado informacion por confidencialidad, a parte de su valor. La matriz de entrada se encuentra en x_train, e y_train contiene la etiqueta relativa a esa transaccion, un 1 indica que si se pago el microcredito y un 0 que no se pago.

In [6]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import GradientBoostingClassifier

In [7]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [10]:
data_file='datos_pago_microcreditos.csv'

pdl= pd.read_csv(data_file)

In [11]:
print(pdl.head())


   payment var1 var2  var3   var4   var5   var6   var7  var8 var9  ...  var21  \
0  Success   qw   hk  3.11  16.06  -4.60  22.34  13.53  1.53   nv  ...   8.94   
1   Denied   qw   rv  3.35  11.18 -18.55   6.68  12.78  6.62   nv  ...  31.02   
2   Denied   qw   zg  4.15  29.19  18.91  16.40   3.67  5.72   ch  ...  23.26   
3  Success   wv   js  6.23  15.70   2.81   4.46   5.13  8.66   ja  ...  29.25   
4  Success   ma   xn  1.28  20.71  14.98  11.19  17.66  1.13   nv  ...   2.19   

   var22  var23  var24  var25  var26  var27 var28  var29  var30  
0 -12.76     ub  12.06   2.46   4.73  -1.72  0.91     ev   8.00  
1  34.76     cz   1.44   9.44  13.56  -2.24  0.24     ev  -2.90  
2   9.50     ri   7.77   8.70  -1.75   5.96  1.91     ev  22.67  
3  -1.53     ri   8.94  19.33  23.73   5.54  0.85     ev  36.31  
4  10.24     ub   8.92   5.48  -0.28   4.01  1.21     ev  11.33  

[5 rows x 31 columns]


In [12]:
print(pdl.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 31 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   payment  30000 non-null  object 
 1   var1     30000 non-null  object 
 2   var2     30000 non-null  object 
 3   var3     30000 non-null  float64
 4   var4     30000 non-null  float64
 5   var5     30000 non-null  float64
 6   var6     30000 non-null  float64
 7   var7     30000 non-null  float64
 8   var8     30000 non-null  float64
 9   var9     30000 non-null  object 
 10  var10    30000 non-null  object 
 11  var11    30000 non-null  object 
 12  var12    30000 non-null  float64
 13  var13    30000 non-null  object 
 14  var14    30000 non-null  float64
 15  var15    30000 non-null  float64
 16  var16    30000 non-null  float64
 17  var17    30000 non-null  object 
 18  var18    30000 non-null  float64
 19  var19    30000 non-null  object 
 20  var20    30000 non-null  float64
 21  var21    300

In [13]:
print(pdl.describe())

               var3          var4          var5          var6          var7  \
count  30000.000000  30000.000000  30000.000000  30000.000000  30000.000000   
mean       3.018848     13.941943     19.981561      8.009453      9.110358   
std        3.004458     14.028019     19.866781      8.013395      8.976009   
min       -8.190000    -43.660000    -55.100000    -23.630000    -27.580000   
25%        0.990000      4.450000      6.520000      2.550000      3.050000   
50%        3.040000     13.920000     19.980000      8.010000      9.170000   
75%        5.040000     23.520000     33.452500     13.430000     15.152500   
max       15.160000     76.780000    107.080000     44.330000     48.390000   

               var8         var12         var14         var15         var16  \
count  30000.000000  30000.000000  30000.000000  30000.000000  30000.000000   
mean       5.993785      7.020662      2.007760     17.118747     10.919426   
std        6.035204      7.004504      2.001967    

In [14]:
pdl["payment"]=np.where(pdl["payment"]=="Success",1,0)

k=pdl.columns

for col in k:
    if pdl[col].dtype=='object':
        temp=pd.get_dummies(pdl[col],drop_first=True,prefix=col)
        pdl=pd.concat([pdl,temp],axis=1)
        pdl.drop([col],axis=1,inplace=True)


pdl.dropna(axis=0,inplace=True)



In [15]:
pdl_train, pdl_test = train_test_split(pdl, test_size = 0.2,random_state=2)

#x_train=pdl_train.drop(["payment"],1)
x_train=pdl_train.drop(columns="payment")
y_train=pdl_train["payment"]

#x_test=pdl_test.drop(["payment"],1)
x_test=pdl_test.drop(columns="payment")
y_test=pdl_test["payment"]

x_train.reset_index(drop=True,inplace=True)
y_train.reset_index(drop=True,inplace=True)

In [16]:
y_train

0        0
1        1
2        0
3        0
4        1
        ..
23995    0
23996    1
23997    0
23998    0
23999    1
Name: payment, Length: 24000, dtype: int32

In [17]:
x_train.head()

var3   var4   var5   var6   var7  var8  var12  var14  var15  var16  ...  \
0  1.02  22.08   8.69  -9.24  21.06  2.21   1.33   2.00   7.62  -5.77  ...   
1  2.09  11.25  23.46  12.11  29.57  4.77   8.86   2.38  29.48  28.14  ...   
2  6.77  25.95   3.95  17.74  19.78  9.02   3.76   0.72  34.00  17.19  ...   
3  2.75  12.47  -4.54  -1.42 -10.32  9.11   8.37   7.91   4.86   3.67  ...   
4  4.47   7.90  31.60   2.53 -12.66  3.64   1.87   5.07  19.04  28.70  ...   

   var23_da  var23_fe  var23_po  var23_qu  var23_ri  var23_sy  var23_tf  \
0         0         1         0         0         0         0         0   
1         0         0         0         0         0         0         1   
2         0         0         0         0         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         0         0         0         1   

   var23_ub  var23_yv  var29_ev  
0         0         0         0  
1         0         0         0  
2         1         0         1  
3         1         0         1  
4         0         0         1  

[5 rows x 62 columns]

# Pregunta 1:
Construye un  clasificador usando K-NN con 3 vecinos y otro usando un arbol de decision. ¿cual produce mejor resultado? ¿que metrica has usado? segun el arbol de decision, ¿que metricas son las mas relevantes?


In [18]:
#Desarrollo del arbol de decision y el clasificador K-NN con 3 vecinos

knn = KNeighborsClassifier(n_neighbors=3)
knn.fit(x_train, y_train)

tree = DecisionTreeClassifier(random_state=2)
tree.fit(x_train, y_train)

DecisionTreeClassifier(random_state=2)

# ¿Cual produce mejor resultado?
El arbol de decision tiene un accuracy mas alto, por lo tanto ha clasificado mejor en el conjunto de prueba en comparacion con K-NN (3 vecinos) 

K-NN 3 vecinos -> 66%

Arbol de decision -> 78%

In [19]:
y_prediccion_knn = knn.predict(x_test)
y_prediccion_tree = tree.predict(x_test)

accuracy_knn = accuracy_score(y_test, y_prediccion_knn)
accuracy_tree = accuracy_score(y_test, y_prediccion_tree)

print("accuracy K-NN: ", accuracy_knn)
print("accuracy Arbol de decision: ", accuracy_tree)

accuracy K-NN:  0.665
accuracy Arbol de decision:  0.7808333333333334


# ¿Que métrica he utilizado?
Se utiliza la metrica accuracy para analizar que tan frecuente acierta el modelo

Dado que el conjunto está moderadamente desbalanceado (no es extremadamente desigual, pero aún relevante), la métrica accuracy puede ser engañosa. Esto se puede observar mediante la distribucion de las clases: 

In [20]:
print("y: ", y_train.value_counts())


y:  0    14993
1     9007
Name: payment, dtype: int64


# segun el arbol de decision, ¿que metricas son las mas relevantes?
Teniendo en cuenta que las metricas son formas de evaluar qué tan bien está funcionando el modelo. Y concluyendo que las metricas mas relevantes en un arbol de decision se refiere a variables del dataset, entonces se tiene que:   

In [21]:
importances = tree.feature_importances_

features = x_train.columns
feature_importance = pd.Series(importances, index=features)

top_features = feature_importance.sort_values(ascending=False).head(10)
print(" las Variables más relevantes segun el arbol de decisiones: ")
print(top_features)

 las Variables más relevantes segun el arbol de decisiones: 
var25       0.359539
var17_bw    0.089330
var16       0.078087
var7        0.051179
var23_ub    0.047099
var22       0.021257
var24       0.020780
var26       0.019466
var5        0.018959
var14       0.018841
dtype: float64


Las variables mas relevantes descritas por el arbol de decisiones son aquellas que mayor influencia tuvieron en las decisiones del modelo, en este caso var25 tiene una mayor importancia con un 35% de influencia en el modelo 

# Pregunta 2:
Repite el ejercicio anterior usando normalizacion de los datos y compara los resultados.

In [22]:
scaler = MinMaxScaler()
x_train_norm = scaler.fit_transform(x_train)
x_test_norm = scaler.fit_transform(x_test)

modelo_norm = DecisionTreeClassifier(random_state=2)
modelo_norm.fit(x_train_norm, y_train)

pred_norm = modelo_norm.predict(x_test_norm)

print("Con los datos normalizados", accuracy_score(y_test, pred_norm))

print(classification_report(y_test, pred_norm))

Con los datos normalizados 0.705
              precision    recall  f1-score   support

           0       0.81      0.70      0.75      3762
           1       0.58      0.72      0.65      2238

    accuracy                           0.70      6000
   macro avg       0.70      0.71      0.70      6000
weighted avg       0.72      0.70      0.71      6000



Aplicando normalizacion a los datos el modelo tiene una exactitud general del 70% de los casos 

En la clase 0 (No pagó el crédito) se puede observar que el modelo acertó el 81%. Además, logró un recall del 70%, es decir, identificó correctamente al 70% de los clientes que realmente no pagaron. Y el promedio entre precision y recall fue del 75% (F1-Score)

En la clase 1 (Sí pagó) el modelo acertó el 58% de las veces. Esto se puede deber al desbalance entre las clases. De los que realmente si pagaron, el modelo acertó en el 72% (Recall) y el promedio entre precision y Recall fue del 65% (F1-Score)   

# Pregunta 3:
Usando GridSearchCV, identifica para un clasificador K-NN el numero de vecinos entre 1 y 30 que optimiza el resultado usando como scoring la precision y con 10 folds para la validacion cruzada. 

In [23]:
param_grid = {'n_neighbors': list(range(1, 31))}

grid_search = GridSearchCV(
    estimator=knn,
    param_grid=param_grid,
    scoring='precision',
    cv=10,
    n_jobs=-1
)

In [24]:
grid_search.fit(x_train_norm, y_train)

print("mejor numero de vecinos: ", grid_search.best_params_['n_neighbors'])
print("Mejor precision promedio en validacion cruzada: ", grid_search.best_score_)

mejor numero de vecinos:  30
Mejor precision promedio en validacion cruzada:  0.6081400121954836


Tras aplicar el clasificador K-NN con un valor óptimo de vecinos igual a 30 (determinado mediante validación cruzada con GridSearchCV optimizando la precisión). El modelo alcanzó una precisión global del 68%.

# Pregunta 4:
Obten la matriz de confusion del clasificador optimo anterior. ¿como lees cada uno de los valores?¿que valor de precission y recall tiene el clasificador optimizado con precission?

In [25]:
knn2 = KNeighborsClassifier(n_neighbors=30)
knn2.fit(x_train, y_train)

y_pred = knn2.predict(x_test)

matriz_confusion = confusion_matrix(y_test, y_pred)
print("Matriz de confusión: \n", matriz_confusion)

reporte = classification_report(y_test, y_pred)
print("\n Reporte de clasificacion: \n", reporte)

Matriz de confusión: 
 [[3480  282]
 [1420  818]]

 Reporte de clasificacion: 
               precision    recall  f1-score   support

           0       0.71      0.93      0.80      3762
           1       0.74      0.37      0.49      2238

    accuracy                           0.72      6000
   macro avg       0.73      0.65      0.65      6000
weighted avg       0.72      0.72      0.69      6000



# matriz de confusion
3480 transacciones fueron correctamente clasificadas como no pagadas (clase 0)

818 transacciones fueron correctamente clasificadas como pagadas (clase 1)

282 transacciones que sí fueron pagadas, el modelo las clasificó como no pagadas (falsos negativos)

1420 transacciones que no fueron pagadas, el modelo las clasificó como pagadas (falsos positivos)

# Reporte de clasificacion
el modelo identifica correctamente la mayoria de los casos en los que no se pagó el crédito 93% (Recall). Sin embargo, tiene dificultades para identificar correctamente a quienes sí pagaron 37% (Recall), aunque cuando lo hace, acierta un 74% (precision). Esto indica un desequilibrio en el rendimiento entre clases, posiblemente al desbalance entre los datos        

# Pregunta 5:
Construye ahora una red neuronal usando MLPClassifier de dos capas. Prueba diferente numero de neuronas por capa. ¿afecta al resultado?

In [26]:
estructura = (100, 50)

In [27]:
mlp = MLPClassifier(hidden_layer_sizes= estructura, max_iter=100, random_state=42)
mlp.fit(x_train, y_train)
pred = mlp.predict(x_test)

c:\Users\ZUN\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(


In [28]:
print("Matriz de confusion: ")
print(confusion_matrix(y_test, pred))
print("\n Reporte de clasificacion: ")
print(classification_report(y_test, pred))

Matriz de confusion: 
[[3074  688]
 [ 457 1781]]

 Reporte de clasificacion: 
              precision    recall  f1-score   support

           0       0.87      0.82      0.84      3762
           1       0.72      0.80      0.76      2238

    accuracy                           0.81      6000
   macro avg       0.80      0.81      0.80      6000
weighted avg       0.81      0.81      0.81      6000



La red neuronal logró un accuracy del 81% lo cual indica un buen rendimiento general. El modelo segun los datos, muestra mejor desempeño en la clase 0 (como los anteriores modelos) con una precision del 87% y un Recall del 82%, mientras que para la clase 1 la precision es del 72% y el recall del 80%. El modelo tiende a clasificar mas correctamente a los clientes que pagan, pero aun conserva una buena capacidad para detectar a los que no lo hacen. La matriz de confucion respalda el analisis, dado que se muestra mas falsos positivos que falsos negativos  

In [30]:
estructurav2 = (150, 100)

In [31]:
mlp = MLPClassifier(hidden_layer_sizes= estructurav2, max_iter=150, random_state=42)
mlp.fit(x_train, y_train)
pred = mlp.predict(x_test)

c:\Users\ZUN\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (150) reached and the optimization hasn't converged yet.
  warnings.warn(


In [32]:
print("Matriz de confusion: ")
print(confusion_matrix(y_test, pred))
print("\n Reporte de clasificacion: ")
print(classification_report(y_test, pred))

Matriz de confusion: 
[[3267  495]
 [ 826 1412]]

 Reporte de clasificacion: 
              precision    recall  f1-score   support

           0       0.80      0.87      0.83      3762
           1       0.74      0.63      0.68      2238

    accuracy                           0.78      6000
   macro avg       0.77      0.75      0.76      6000
weighted avg       0.78      0.78      0.78      6000



Al aumentar las capas ocultas y el número de iteraciones. La precision en la clase 1 aumentó de 72% a 74%, mientras que la precision en la clase 0 bajo considerablemente de 87% a 80%, se puede observar que Recall se mantiene relativamente estable, lo que indica que el modelo sigue teniendo dificultad para identificar todos los casos. F1-Score para clase 1 mejoró un poco, pasando de 66% a 68%, lo cual se interpreta un mejor equilibrio entre precision y Recall  

### Pregunta 6:
GradientBoostingClasifier es uno de los metodos de scikitlearn que mejor resultados suelen producir. Implementa un clasificador utilizando GradientBoostingClasifier considerando 0.05, 0.1, 0.15, 0.2, 0.25, 0.3 como learning rate.¿que learning rate produce un resultado optimo usando recall?

In [33]:
learning_rate = [0.05, 0.1, 0.15, 0.2, 0.25, 0.3]
resultados = []

In [34]:
for lr in learning_rate:
    clf = GradientBoostingClassifier(learning_rate=lr, random_state=42)
    clf.fit(x_train, y_train)
    pred = clf.predict(x_test)

    report = classification_report(y_test, pred, output_dict=True)
    recall_clase_1 = report['1']['recall']

    resultados.append((lr, recall_clase_1))
    print(f"Learning rate: {lr} | Recall clase 1: {recall_clase_1:.4f}")

Learning rate: 0.05 | Recall clase 1: 0.7895
Learning rate: 0.1 | Recall clase 1: 0.8061
Learning rate: 0.15 | Recall clase 1: 0.8052
Learning rate: 0.2 | Recall clase 1: 0.8065
Learning rate: 0.25 | Recall clase 1: 0.8061
Learning rate: 0.3 | Recall clase 1: 0.8052


El mejor learning rate en terminos de recall para la clase 1 es 0.2, con un valor de 0.80 

### Pregunta 6:
Usando la matriz de confusion de los tres casos (GradientBoosting, MLP y K-NN) y derivando las metricas que te parezcan oportunas ¿Que clasificador de los 3 seleccionarias y porque?

In [49]:
def evaluar_modelo(nombre, y_test, y_pred):
    cm = confusion_matrix(y_test, y_pred)
    report = classification_report(y_test, y_pred, output_dict=True)
    accuracy = accuracy_score(y_test, y_pred)
    recall_clase1 = report['1']['recall']
    f1_clase1 = report['1']['f1-score']
    print(f"\n Modelo: {nombre}")
    print(f"Matriz de confusion: \n {cm}")
    print(f"Recall clase 1: {recall_clase1:.4f}")
    print(f"f1_score clase 1: {f1_clase1:.4f}")
    print(f"Accuracy: {accuracy:.4f}")
    return{
        "modelo": nombre,
        "recall": recall_clase1,
        "f1": f1_clase1,
        "Accuracy": accuracy
    }

In [47]:
resultados = []

gb = GradientBoostingClassifier(learning_rate=0.2)
gb.fit(x_train, y_train)
y_pred_gb = gb.predict(x_test)
resultados.append(evaluar_modelo("Gradient Boosting", y_test, y_pred_gb))



 Modelo: Gradient Boosting
Matriz de confusion: 
 [[3407  355]
 [ 433 1805]]
Recall clase 1: 0.8065
f1_score clase 1: 0.8208
Accuracy: 0.8687


# Modelo: Gradient Boosting
Es un metodo de aprendizaje supervisado que combina multiples arboles de decisiones pequeños para crear un modelo que intenta corregir los errores de los anteriores. 

Matriz de confusion: 
3407 casos donde el modelo predijo 0 y efectivamente eran 0

355 casos donde el modelo predijo 1, pero en realidad era 0

433 casos donde el modelo predijo 0, pero en realidad era 1

1805 casos donde el modelo predijo 1 y efectivamente eran 1

El modelo identifico correctamente el 80.6% de todos lo casos positivos (Recall clase 1)

El modelo tiene una media armonica entre presicion y recall del 82% (f1_score clase1)

El modelo mantiene una presicion del 86.8% (Accuracy)

In [48]:
resultados.append(evaluar_modelo("MLP", y_test, pred))


 Modelo: MLP
Matriz de confusion: 
 [[3397  365]
 [ 436 1802]]
Recall clase 1: 0.8052
f1_score clase 1: 0.8182
Accuracy: 0.8665


# Modelo: MLP
Es un metodo de aprendizaje. 

Matriz de confusion:

3397 casos donde el modelo prefijo 0 y efectivamente eran 0

365 casos donde el modelo predijo 1, pero en realidad era 0

436 casos donde el modelo predijo 0, pero en realidad era 1

1802 casos donde el modelo predijo 1 y efectivamente si es 1

El modelo identifico correctamente el 80.5% de todos lo casos positivos (Recall clase 1)

El modelo tiene una media armonica entre presicion y recall del 81.8% (f1_score clase1)

El modelo mantiene una presicion del 86.6% (Accuracy)

In [50]:
resultados.append(evaluar_modelo("K-NN", y_test, y_prediccion_knn))


 Modelo: K-NN
Matriz de confusion: 
 [[2901  861]
 [1149 1089]]
Recall clase 1: 0.4866
f1_score clase 1: 0.5201
Accuracy: 0.6650


# Modelo: k-NN
Es un metodo de aprendizaje. 

Matriz de confusion:     
2901 casos en el que el modelo predijo exitosamente que era 0

861 casos en el que el modelo predijo que era 1, pero en realidad era 0

1149 casos en el que el modelo predijo 0, pero en realidad era 1

1089 casos en el que el modelo predijo 1 y exitosamente si era 1

El modelo identifico correctamente el 48.6% de todos lo casos positivos (Recall clase 1)

El modelo tiene una media armonica entre presicion y recall del 52% (f1_score clase1)

El modelo mantiene una presicion del 66.5% (Accuracy)


# Conclusion 
De los tres clasificadores seleccionaria el modelo Gradient Boosting, debido a que es ligeramente superior en todas las metricas 